src/policy/semantic.rs

```diff
@@ -286,36 +256,86 @@ impl<Pk: MiniscriptKey> Policy<Pk> {
-                let mut ret_subs = Vec::with_capacity(subs.len());
-                for sub in subs {
-                    match sub.normalized() {
-                        Policy::Trivial => {}
-                        Policy::Unsatisfiable => return Policy::Unsatisfiable,
-                        Policy::And(and_subs) => ret_subs.extend(and_subs),
-                        x => ret_subs.push(x),
+            Policy::Threshold(k, subs) => {
+                if k == subs.len() {
+                    let mut ret_subs = Vec::with_capacity(subs.len());
+                    for sub in subs.clone() {
+                        match sub.normalized() {
+                            Policy::Trivial => {}
+                            Policy::Unsatisfiable => return Policy::Unsatisfiable,
+                            Policy::Threshold(k, and_subs) => {
+                                if k == and_subs.len() {
+                                    ret_subs.extend(and_subs);
+                                } else {
+                                    ret_subs.push(Policy::Threshold(k, and_subs));
+                                }
+                            }
+                            x => ret_subs.push(x),
+                        }
                     }
-                }
-                match ret_subs.len() {
-                    0 => Policy::Trivial,
-                    1 => ret_subs.pop().unwrap(),
-                    _ => Policy::And(ret_subs),
-                }
-            }
-            Policy::Or(subs) => {
-                let mut ret_subs = Vec::with_capacity(subs.len());
-                for sub in subs {
-                    match sub {
-                        Policy::Trivial => return Policy::Trivial,
-                        Policy::Unsatisfiable => {}
-                        Policy::Or(or_subs) => ret_subs.extend(or_subs),
-                        x => ret_subs.push(x),
+                    match ret_subs.len() {
+                        0 => Policy::Trivial,
+                        1 => ret_subs.pop().unwrap(),
+                        k => Policy::Threshold(k, ret_subs),
+                    }
+                } else if k == 1 {
+                    let mut ret_subs = Vec::with_capacity(subs.len());
+                    for sub in subs {
+                        match sub.normalized() {
+                            Policy::Trivial => return Policy::Trivial,
+                            Policy::Unsatisfiable => {}
+                            Policy::Threshold(1, or_subs) => {
+                                ret_subs.extend(or_subs);
+                            }
+                            x => ret_subs.push(x),
+                        }
+                    }
+                    match ret_subs.len() {
+                        0 => Policy::Unsatisfiable,
+                        1 => ret_subs.pop().unwrap(),
+                        _ => Policy::Threshold(1, ret_subs),
+                    }
+                } else {
```

Comment:
Do we actually need the `and` and `or` branches or can we just use the general one always?

Reply:
We do a recursive combination of branches. `thresh(1,pk(A),thresh(1,pk(B),pk(C)))` would be reread as `thresh(1,pk(A),pk(B),pk(C))`

Reply:
Actually, you may be right. Thinking a bit more about this

Reply:
You are right, all of them can be covered in one general condition. I blindly translated the `And` and `or` thinking that those were necessary for merging